pkg/distributors/contracts/MultiDistributor.sol

```diff
@@ -494,134 +477,169 @@ contract MultiRewards is IMultiRewards, IDistributor, ReentrancyGuard, MultiRewa
-    }
+            if (userStaking.subscribedDistributions.contains(distributionId)) {
+                // Update user distribution rates only if the user is still subscribed
+                _updateUserRewardRatePerToken(userStaking, distributionId);
+            }
 
-    /**
-     * @notice Set the reward duration for a reward
-     * @param stakingToken The staking token to be set
-     * @param rewardsToken The token for the reward
-     * @param duration The duration over which each distribution is spread
-     */
-    function setRewardsDuration(
-        IERC20 stakingToken,
-        IERC20 rewardsToken,
-        uint256 duration
-    ) external {
-        require(duration > 0, "Reward duration must be non-zero");
+            UserDistribution storage userDistribution = userStaking.distributions[distributionId];
+            uint256 unpaidRewards = userDistribution.unpaidRewards;
 
-        bytes32 distributionId = getDistributionId(stakingToken, rewardsToken, msg.sender);
-        Distribution storage distribution = _getDistribution(distributionId);
-        require(distribution.duration > 0, "Reward must be configured with addReward");
-        require(distribution.periodFinish < block.timestamp, "Reward period still active");
+            if (unpaidRewards > 0) {
+                userDistribution.unpaidRewards = 0;
+               
+                IAsset rewardsToken = IAsset(address(distribution.rewardsToken));
+                // Iterate through all the tokens we've seen so far.
+                for (uint256 j; j < tokens.length; j++) {
+                    // Check if we're already sending some of this token
+                    // If so we just want to add to the existing transfer
+                    if (tokens[j] == rewardsToken) {
+                        amounts[j] += unpaidRewards;
+                        break;
+                    } else if (tokens[j] == IAsset(0)) {
+                        // If it's the first time we've seen this token
+                        // record both its address and amount to transfer
+                        tokens[j] = rewardsToken;
+                        amounts[j] = unpaidRewards;
+                        numTokens += 1;
+                        break;
+                    }
+                }
+            }
+        }
 
-        distribution.duration = duration;
-        emit RewardDurationSet(distributionId, duration);
+        IVault.UserBalanceOp[] memory ops = new IVault.UserBalanceOp[](numTokens);
+        for (uint256 i; i < numTokens; i++) {
+            ops[i] = IVault.UserBalanceOp({
+                asset: tokens[i],
+                amount: amounts[i],
+                sender: address(this),
+                recipient: payable(recipient),
+                kind: kind
+            });
+
+            if (amounts[i] > 0){
+                emit RewardPaid(msg.sender, address(tokens[i]), amounts[i]);
+            }
+        }
+
+        getVault().manageUserBalance(ops);
```

Comment:
I've included the mechanism of batching transfers used in #827. This seems more suitable to me than the method used in #894 as a simplified interface is more important here as `MultiDistributor` is likely to have many contracts as users.